python/src/main/python/yaml-template/main.py

```diff
@@ -13,18 +13,52 @@
 # License for the specific language governing permissions and limitations under
 # the License.
 #
-
+#
 import argparse
+import json
 import logging
+import os
+from argparse import Namespace
+
+import jinja2
 
+from apache_beam.io.filesystems import FileSystems
 from apache_beam.yaml import cache_provider_artifacts
 from apache_beam.yaml import main
 
 
-def run(argv=None):
+def _configure_parser(argv):
   parser = argparse.ArgumentParser()
-  _, pipeline_args = parser.parse_known_args(argv)
-  pipeline_args += ['--sdk_location=container']
+  parser.add_argument(
+    '--jinja_variables',
+    default=None,
+    type=json.loads,
+    help='A json dict of variables used when invoking the jinja preprocessor '
+         'on the provided yaml pipeline.')
+  return parser.parse_known_args(argv)
+
+
+class _BeamFileIOLoader(jinja2.BaseLoader):
+  def get_source(self, environment, path):
+    source = FileSystems.open(path).read().decode()
+    return source, path, lambda: True
+
+
+def run(argv=None):
+  known_args, pipeline_args = _configure_parser(argv)
+  known_args = Namespace(
+    **vars(main._configure_parser(argv)[0]), **vars(known_args))
+  pipeline_yaml = (
+    jinja2.Environment(
+      undefined=jinja2.StrictUndefined, loader=_BeamFileIOLoader())
+    .from_string(main._pipeline_spec_from_args(known_args))
+    .render(**known_args.jinja_variables or {}))
+  pipeline_yaml = os.linesep.join(
+    [s for s in pipeline_yaml.splitlines() if s.strip()])
+
+  pipeline_args += [
+    '--sdk_location=container', f'--yaml_pipeline={pipeline_yaml}'
```

Comment:
Won't we need to remove any `--yaml_pipeline` or --yaml_pipeline_file` arguments as well as append this one?

Reply:
We're not passing through `known_args` so I think this is fine, though I'm not actually sure why that is now that I think of it

Reply:
+1
Because the original ones are already getting processed so we don't need to reprocess

Reply:
Actually, I did not consider the `--json_schema_validation` flag... maybe I should refactor to just omit the original `yaml_pipeline` and `yaml_pipeline_file` flags